spec/index.bs

```diff
@@ -12071,77 +12071,131 @@ partial interface GPUDevice {
-    : <dfn>\[[error]]</dfn>, of type {{GPUError}}?, initially `null`
-    ::
-        The first {{GPUError}}, if any, observed while the [=GPU error scope=] was current.
+{{GPUError}} is the base interface for all errors surfaced from {{GPUDevice/popErrorScope()}}
+and the {{GPUDevice/uncapturederror}} event.
 
-    : <dfn>\[[filter]]</dfn>, of type {{GPUErrorFilter}}
+{{GPUError}} has the following attributes:
+
+<dl dfn-type=attribute dfn-for=GPUError>
+    : <dfn>message</dfn>
     ::
-        Determines what type of {{GPUError}} this [=GPU error scope=] observes.
+        A human-readable message providing information about the error that occurred.
+
+        Note: This message is generally intended for application developers to debug their
+        applications and capture information for debug reports, not to be surfaced to end-users.
+
+        Note: User agents should not include potentially machine-parsable details in this message,
+        such as free system memory on {{GPUErrorFilter/"internal"}} errors raised due to memory
+        constraints, or other details about the conditions under which memory was exhausted.
 </dl>
 
 <script type=idl>
-enum GPUErrorFilter {
-    "out-of-memory",
-    "validation"
+[Exposed=(Window, DedicatedWorker), SecureContext]
+interface GPUValidationError : GPUError {
+    constructor(DOMString message);
 };
 </script>
 
+{{GPUValidationError}} is a subtype of {{GPUError}} which indicates that an operation did not
+satisfy all validation requirements. Validation errors are always indicative of an application
+error, and is expected to fail the same way across all devices assuming the same
+{{device/[[features]]}} and {{device/[[limits]]}} are in use.
+
 <script type=idl>
-[Exposed=(Window, DedicatedWorker), SecureContext]
-interface GPUError {
-    readonly attribute DOMString message;
+enum GPUInternalErrorReason {
+    "out-of-memory"
 };
 
 [Exposed=(Window, DedicatedWorker), SecureContext]
-interface GPUOutOfMemoryError : GPUError {
-    constructor(DOMString message);
-};
+interface GPUInternalError : GPUError {
+    constructor(DOMString message, optional GPUInternalErrorReason reason);
 
-[Exposed=(Window, DedicatedWorker), SecureContext]
-interface GPUValidationError : GPUError {
-    constructor(DOMString message);
+    readonly attribute (GPUInternalErrorReason or undefined) reason;
 };
 </script>
 
-{{GPUError}} is the base interface for all errors surfaced from {{GPUDevice/popErrorScope()}}
-and the {{GPUDevice/uncapturederror}} event.
+{{GPUInternalError}} is a subtype of {{GPUError}} which indicates than an operation failed for a
+system or implementation-specific reason even when all validation requirements have been satisfied.
+For example, the operation may exceede the capabilities of the implementation in a way not easily
+captured by the [=supported limits=]. The same operation may succeed on other devices or under
+difference circumstances.
```

Comment:
We should say that internal errors can only come from certain operations described in the spec, not any arbitrary operation.

Reply:
Done, and also added an algorithm for generating an internal error for future use (and while I was at it cleaned up some of the algorithms to use the new timeline styles.)

Haven't actually added any instances of an operation generating an internal error yet, because this CL is already messy enough as is and I'd like those to be more easily reviewed.

Out of curiosity, what's the expected behavior for an operation that we don't explicitly say can generate internal error experiencing an internal error? Is that just forced to be a lost context every time?

Reply:
I didn't realize we didn't currently specify that out of memory can come from createBuffer/Texture/QuerySet. That'll be nice to have written down.

Yeah, the behavior I have in mind is:
- Implementation does any heroics it wants to try to do to make the error go away (like waiting for the queue to finish and then deleting any temporary resources) and tries again
- If it still fails, lose the device